[cxx-interop] Allow C++ function templates to be instantiated with Swift closures

j-hui opened the pull request:
Swift started to explicitly forbid the instantiation of C++ function templates with arbitrary types in #77430, because many types cause the Swift compiler to crash. However, those checks prevented them from being instantiated with Swift closures (which were previously fine), causing a regression.

This patch relaxes the convertTemplateArgument() function to also allow converting Swift function types, and adds some tests to make sure doing so is fine. This patch also does some cleanup of an existing test checking the instantiation of various types.

rdar://148124104

j-hui committed:
[cxx-interop] Allow C++ function templates to be instantiated with Swift closures

Swift started to explicitly forbid the instantiation of C++ function
templates with arbitrary types in #77430, because many types cause the
Swift compiler to crash. However, those checks prevented them from being
instantiated with Swift closures (which were previously fine), causing
a regression.

This patch relaxes the convertTemplateArgument() function to also allow
converting Swift function types, and adds some tests to make sure doing
so is fine.

This patch also does some cleanup of existing tests checking the
instantiation of various types, and adds testing for C function pointers
and Obj-C blocks.

rdar://148124104

j-hui commented:
@swift-ci please test

Xazax-hun commented on test/Interop/Cxx/templates/function-template.swift:
```diff
@@ -39,6 +39,97 @@ FunctionTemplateTestSuite.test("constLvalueReferenceToBool<T> where T == Bool")
+  expectEqual(value, 222)
+}
+
+FunctionTemplateTestSuite.test("indirectlyCallFunction<T> where T == () -> ()") {
```
I am not sure how blocks are handled but is it possible to import a block from Obj-C? If so, is that imported as a closure? Could we instantiate a C++ template with a block that we imported from Obj-C?

j-hui replied:
I _think_ that should be possible. I'll add a test for that.

j-hui replied:
Something that _doesn't_ work yet, though, are C function pointers, because pointers in general are still quite broken.

j-hui replied:
So it turns out C function pointers work; I've updated the test suite to cover those too. It's only _optional_ (function) pointers that are broken.

j-hui commented on lib/AST/ClangTypeConverter.cpp:
```diff
@@ -991,6 +1007,22 @@ clang::QualType ClangTypeConverter::convertTemplateArgument(Type type) {
   }
 
+  if (auto functionType = type->getAs<FunctionType>()) {
+    if (llvm::none_of(functionType->getParams(), [](auto &param) {
```
Oh, good catch! I'll add this check to SILFunctionType as well.

j-hui replied:
I actually decided to remove the check. _All_ template instantiations involving FRTs are wonky because of nullable<->optional round-tripping, so I think we should keep that behavior at least consistent (even if the behavior is wrong).

j-hui replied:
(specifically, the behavior is that the compiler crashes...)

j-hui committed:
[NFC] Make templateArgument a template argument

This Boolean flag is used in ClangTypeConverter to indicate whether
a type is being converted in the context of a template type parameter.
This parameter can be made a template parameter because it is always
a compile-time constant.

j-hui commented on lib/AST/ClangTypeConverter.cpp:
```diff
@@ -165,16 +167,19 @@ const clang::Type *ClangTypeConverter::getFunctionType(
-                     : ClangASTContext.VoidTy;
+const clang::Type *ClangTypeConverter::getFunctionType(
+    ArrayRef<SILParameterInfo> params, std::optional<SILResultInfo> result,
+    SILFunctionType::Representation repr, bool templateArgument) {
```
I've done this in a separate `[NFC]` patch (also part of this PR), which I do not plan on squashing with the main `[cxx-interop]` patch when I merge (I shall make a merge commit).

j-hui commented:
@swift-ci please test

j-hui commented:
Hm. Apparently raw C function pointers don't work without the blocks runtime either. I'm going to move these to the `// REQUIRES: objc_runtime` test for now.

j-hui commented:
@swift-ci please test

Xazax-hun reviewed (approved):
Thanks!

Xazax-hun commented:
@swift-ci please test macOS

j-hui commented:
@swift-ci please test macOS

j-hui commented:
@swift-ci please test macos

j-hui commented:
@swift-ci please test

j-hui commented:
@swift-ci please test macos
